sdk/program/PROJECT.md

```diff
@@ -0,0 +1,138 @@
+
+## Project Plan Phase 1: solana_program updates
+
+### Break it up: if it can reasonably be used independently, it gets its own crate
+
+As one approach to address point number 4 of easily allowing customization, we
+should avoid bundling too much functionality in solana-program, and instead
+break it up sensibly. This way, someone can easily drop in a custom entrypoint
+without pulling in the default during build time.
+
+This also helps address point number 8 of avoiding kitchen sinks. By putting
+only related functionality into separate crates, there is no temptation for
+developers to "just throw it into the SDK." Instead, they'll have to consider
+the best place for their changes.
+
+When breaking up solana-program, we will take it in phases. First, we'll extract
+functionality required for program clients, which typically involve creating
+instructions and deserializing types. Next, we will actually create program
+client crates for the native programs. And finally, we will extract functionality
+used for writing on-chain programs.
+
+Here's a more specific breakdown:
+
+* Add crates required for program clients:
+  * borsh
+  * clock
+  * example-mocks
+  * hash
+  * instruction
+  * message
+  * program-error
+  * pubkey
+  * serde-varint
+  * sysvar
+* Those crates allow us to make program client crates:
+  * address lookup table
+  * bpf loader upgradeable
+  * feature
+  * loader v4
+  * stake
+  * system
+  * vote
+* Next, focus on crates for writing programs:
+  * account-info
+  * entrypoint
+  * program-pack
+  * program-syscalls
+  * stable-layout
+* Finally, do all the rest:
+  * native-token
+  * serialize-utils
+  * wasm
+* Deprecate any bits that make sense
+* `solana_program` becomes a crate with only re-exports and deprecated code
+
+### Add crate features
+
+After solana-program has been broken up, the next bit is to address point number
+7 and minimize dependencies. This will be done 
+
+* Serialization: serde / borsh / bincode / bytemuck optional
+* Syscalls: `#[cfg(not(target_os = "solana"))]` implementations of syscalls (ie.
+`hashv`) pull in unneeded dependencies. Where possible, either gate these deps
+with `[target.'cfg(not(target_os = "solana"))'.dependencies]` or create features
+like `backup-syscall`
+* Add an `unstable` feature for new code that could break, similar to Rust
+
+### Update CI
+
+* Reuse scripts and workflow for publishing crates:
+  * `workflows/release.yml`
+  * `publish-crate.sh`, minus dockerization
+* Stop using Buildkite, use only GitHub Actions for these steps:
+  * `test-sanity`
+  * `shellcheck`
+  * `test-checks`
+  * `test-coverage`
+  * `test-stable`
+  * `test-wasm`
+  * `test-docs`
+* Add downstream testing (along with SPL and Anchor):
+  * Agave monorepo clippy
+  * Template programs
+  * New program from scratch
+* Backport
+  * The new repo uses the same branching model as the monorepo, and contains
+    a v2.0 branch, but not v1.18
+
+### Move solana-program to a new repo
```

Comment:
Do we have a particular target date in mind? Can this be done whenever, or do we need to wait for a release?

Reply:
I don't think there's a particular target date, I'll just coordinate with devops whenever we're ready